File: LeetCode/Easy/1.TwoSum/C/TwoSum.c
// Began at around 2:30 at night.
// Stopped at 3:30(I'm tired, also programming in C, from ground-up is time-consuming :( ).

#include <stdio.h>
#include <stdlib.h>

int* twoSum(int* nums, int numsSize, int target, int* returnSize) {

    int* result = (int*)malloc(sizeof(int) * 2);
    *returnSize = 0;

    for (int i = 0;i < numsSize - 1;i++)
    {
        for (int j = i + 1;j < numsSize;j++)
        {
            if (nums[j] == target - nums[i])
            {
                result[0] = i;
                result[1] = j;
                *returnSize = 2;
                return result;
            }
        }
    }

    free(result);  // we free this if incase no solution is found
    return NULL;  // here, we couldn't return -1 since the return type of this function is int*, that is, a pointer.
}

int main()
{
    int size,target,returnSize;
    do
    {
        printf("Enter the size of the Array : ");
        scanf("%d",&size);
    } while (size < 2);  // apparently, out of all the loops in C, just the do-while loop ends with a semi-colon. Also, the min-length is 2 because it's given in the problem's constraints

    int* array1 = (int*)malloc(sizeof(int) * size);
    
    for (int i = 0;i < size;i++)
    {
        printf("Enter the %dth element of the array : ",i + 1);
        scanf("%d",&array1[i]);
    }

        printf("Enter the target element : ");
        scanf("%d",&target);

    int* result = twoSum(array1,size,target,&returnSize);

    if (returnSize != 2)
    {
        printf("The array does not contain two elements that sum up to the target :(\n");  // we used \n newline character so that the cmd line appears on the next line, otherwise it appears on the same line at the end.
    } else
    {
        printf("The indices of the elements of the array elements which add up to the target(%d) element are: %d and %d \n",target,result[0],result[1]);  // we used \n newline character so that the cmd line appears on the next line, otherwise it appears on the same line at the end.
        free(result);
    }

    free(array1);
    return 0;
}

/*
    I initially had a conditional statement at the end of the end of the do-while loop at lOC 37 `while (size < 2 || typeof(size) != int)`,
    but then I realized that I need not check the type of 'size' variable since that job is done by 'scanf' function. 
    I had also wrapped LOCs 43 and 44 in do-while (type != int); but then later I rectified it but removing it.
*/

File: LeetCode/Easy/1.TwoSum/Typescript/TwoSum.ts
// method 1 to try to make user-input through 'prompt()' work in Typescript
// const prompt = require('prompt-sync')();

// method 2
// import promptSync from 'prompt-sync';
// const prompt = promptSync();

// sadly, I couldn't find a way to replicate the solution which I used in the JS version of this Leetcode problem.

// Nevermind, this works. I had to run `npm run prompt-sync` in the terminal.   (is it npm install prompt-sync /or/ npm run prompt-sync ?)

import * as PromptSync from 'prompt-sync'

const prompt = PromptSync();

function TwoSum(arr1 : number[],target : number) : number | number[] {
    for (let i = 0;i < arr1.length - 1;i++) {
        for (let j = 0;j < arr1.length;j++) {
            if (arr1[j] == target - arr1[i]) {
                return [i + 1,j + 1];
            }
        }
    }

    return (-1);
}

function main() {
    // var array1 : number[] = [1,2,3,4,5];
    // var target : number = 9;

    var size : number = Number(prompt("Enter the size of the array : "));
    var array1 : number[] = [];

    for (let i = 0;i < size;i++) {
        let a : number = Number(prompt(`Enter the ${i + 1}th element : `));
        array1.push(a);
    }

    var target : number = Number(prompt("Enter the target element : "));

    var result : number | number[]= TwoSum(array1,target);

    if (result == (-1)) {
        console.log("No two array elements add up to the target");
    } else {
        console.log(`The indices of the elements which add up to the target element(${target}) are : ${result[0]} and ${result[1]}`);
    }
}

main();